src/fastalloc/iter.rs

```diff
@@ -0,0 +1,643 @@
+use crate::{Operand, OperandKind, OperandPos, OperandConstraint};
+
+#[derive(Clone, Copy, PartialEq)]
+enum OperandConstraintKind {
+    Any,
+    Reg,
+    Stack,
+    FixedReg,
+    Reuse,
+}
+
+impl PartialEq<OperandConstraint> for OperandConstraintKind {
+    fn eq(&self, other: &OperandConstraint) -> bool {
+        match other {
+            OperandConstraint::Any => *self == Self::Any,
+            OperandConstraint::Reg => *self == Self::Reg,
+            OperandConstraint::Stack => *self == Self::Stack,
+            OperandConstraint::FixedReg(_) => *self == Self::FixedReg,
+            OperandConstraint::Reuse(_) => *self == Self::Reuse,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+struct SearchConstraint {
```

Comment:
This is a neat abstraction and is certainly providing some benefit by generalizing all the cases below. However I worry a little bit about efficiency here: we're explicitly building a data structure in memory to define what kind of operands we're matching, and then dispatching based on this (iterating, in some cases). Data-driven control flow that is "constant in practice" can sometimes be optimized but often not. It may be simpler and faster to do something like:

```
impl<'a> Operands<'a> {
  fn matching<F: Fn(Operand)>(&self, predicate: F) -> impl Iterator<Item = Operand> + 'a {
    self.iter().filter(|op| predicate(*op))
  }
}
```

then

```
impl<'a> Operands<'a> {
  fn non_fixed_non_late(&self) -> impl Iterator<Item = Operand> + 'a {
    self.matching(|op| op.pos() == ... && op.constraint().kind() != ...))
  }
}
```

This offers a little more flexibility (vs. e.g. the need to have two slots for "must not have") and should inline nicely into a filtered for-loop.